### PR TITLE
[#1801] Using constructor as builderCreationMethod in custom builder …

### LINKED ISSUE #1801: Using constructor as builderCreationMethod in custom builder provider causes NullPointerException
I am building a custom builder provider and using constructor as the builderCreationMethod.

Something like

```java
        List<TypeElement> types = ElementFilter.typesIn(typeElement.getEnclosedElements());
        List<BuilderInfo> builderInfo = new ArrayList<>();
        types.stream()
             .filter(t -> !this.typeUtils.isSameType(t.asType(), typeElement.asType()))
             .filter(innerType -> isPossibleInnerBuilder(innerType, typeElement))
             .forEach(innerBuilderClass -> {
                 Collection<ExecutableElement> buildMethods = findBuildMethods(innerBuilderClass, typeElement);
                 if (!buildMethods.isEmpty()) {
                     ElementFilter.constructorsIn(innerBuilderClass.getEnclosedElements())
                                  .stream()
                                  .filter(constructor -> constructor.getParameters().isEmpty())
                                  .findAny()
                                  .ifPresent(defaultConstructor -> builderInfo.add(
                                      new BuilderInfo.Builder()
                                          .builderCreationMethod(defaultConstructor)
                                          .buildMethod(buildMethods)
                                          .build()
                                  ));
                 }
             });
```

It will cause NPE during code generation

```java
error: Internal error in the mapping processor: java.lang.NullPointerException

at org.mapstruct.ap.internal.util.workarounds.SpecificCompilerWorkarounds.replaceTypeElementIfNecessary(SpecificCompilerWorkarounds.java:99)    
at org.mapstruct.ap.internal.util.Executables.getAllEnclosedAccessors(Executables.java:136)     
at org.mapstruct.ap.internal.model.common.Type.getAllAccessors(Type.java:647)   
at org.mapstruct.ap.internal.model.common.Type.getPropertyReadAccessors(Type.java:508)          
at org.mapstruct.ap.internal.model.source.TargetReference$BuilderFromTargetMapping.getTargetEntries(TargetReference.java:189)   
at org.mapstruct.ap.internal.model.source.TargetReference$BuilderFromTargetMapping.build(TargetReference.java:158)      
at org.mapstruct.ap.internal.model.source.Mapping.init(Mapping.java:446)        
at org.mapstruct.ap.internal.model.source.Mapping.init(Mapping.java:413)        
at org.mapstruct.ap.internal.model.source.SourceMethod$Builder.build(SourceMethod.java:197) 
```

I did a very brief investigation, the direct cause is in the `BuilderType::create` method when the `builder` in `BuilderType` is initialized. For constructor, the return type is `void`. I think it should use the `owner type` when the `builderCreationMethod` is a constructor. 

```java
    public static BuilderType create(BuilderInfo builderInfo, Type typeToBuild, TypeFactory typeFactory,
        Types typeUtils) {
        if ( builderInfo == null ) {
            return null;
        }

        // The following line won't work for constructor method because the return type of a constructor is 'void'. 
        Type builder = typeFactory.getType( builderInfo.getBuilderCreationMethod().getReturnType() );

        ExecutableElement builderCreationMethod = builderInfo.getBuilderCreationMethod();
        Type owner;
        TypeMirror builderCreationOwner = builderCreationMethod.getEnclosingElement().asType();
        if ( typeUtils.isSameType( builderCreationOwner, typeToBuild.getTypeMirror() ) ) {
```

### PR DESCRIPTION
**Pull request by dengzhizhi**
The return type of constructor is Void. In this case the builder type should be the owner type of the constructor.

Tested with my customized builder provider.

Fixes #1801

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java [hunk 1]
```diff
@@ -100,8 +100,8 @@ public String implementationPackage() {
     public List<TypeMirror> imports() {
         List<TypeMirror> imports = new ArrayList<>();
         imports.addAll( mapperPrism.imports() );
-        if ( mapperConfigPrism != null ) {
-            imports.addAll( mapperConfigPrism.imports() );
+        if ( mapperConfigPrism != null && mapperConfigPrism.imports() != null) {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java]: Are you sure that `mapperConfigPrism.imports()` can return `null`? We don't do a `null` check on line 102 which is the same logic.

**Comment by sjaakd** [comment]
seems that the current code base works different.. I'll check it out.

**Comment by sjaakd** [comment]
@dengzhizhi can I add you to the copyright list? I just rebased this branch and opened a new PR..

**Comment by filiphr** [comment]
Thanks for the PR @dengzhizhi. We are closing this one in favour of #1905 which resolves the conflicts

**Pull request closed without merging**